core/trie/proof.go

```diff
@@ -165,12 +189,328 @@ func VerifyProof(root *felt.Felt, key *Key, value *felt.Felt, proofs []ProofNode
+			if err != nil {
+				return false, err
+			}
+		}
+	}
+
+	// Step 2: Build trie from proofPaths and keys
+	tmpTrie, err := BuildTrie(proofPaths[0], proofPaths[1], keys, values)
+	if err != nil {
+		return false, err
+	}
+
+	// Verify that the recomputed root hash matches the provided root hash
+	recomputedRoot, err := tmpTrie.Root()
+	if err != nil {
+		return false, err
+	}
+	if !recomputedRoot.Equal(root) {
+		return false, errors.New("root hash mismatch")
+	}
+
+	return true, nil
+}
+
+func ensureMonotonicIncreasing(proofKeys [2]*Key, keys []*felt.Felt) error {
+	if proofKeys[0] != nil {
+		leftProofFelt := proofKeys[0].Felt()
+		if leftProofFelt.Cmp(keys[0]) >= 0 {
+			return errors.New("range is not monotonically increasing")
+		}
+	}
+	if proofKeys[1] != nil {
+		rightProofFelt := proofKeys[1].Felt()
+		if keys[len(keys)-1].Cmp(&rightProofFelt) >= 0 {
+			return errors.New("range is not monotonically increasing")
+		}
+	}
+	if len(keys) >= 2 {
+		for i := 0; i < len(keys)-1; i++ {
+			if keys[i].Cmp(keys[i+1]) >= 0 {
+				return errors.New("range is not monotonically increasing")
+			}
+		}
+	}
+	return nil
+}
+
+// compressNode determines if the node needs compressed, and if so, the len needed to arrive at the next key
+func compressNode(idx int, proofNodes []ProofNode, hashF hashFunc) (int, uint8, error) {
+	parent := &proofNodes[idx]
+
+	if idx == len(proofNodes)-1 {
+		if parent.Edge != nil {
+			return 1, parent.Len(), nil
+		}
+		return 0, parent.Len(), nil
+	}
+
+	child := &proofNodes[idx+1]
+
+	switch {
+	case parent.Edge != nil && child.Binary != nil:
+		return 1, parent.Edge.Path.len, nil
+	case parent.Binary != nil && child.Edge != nil:
+		childHash := child.Hash(hashF)
+		if parent.Binary.LeftHash.Equal(childHash) || parent.Binary.RightHash.Equal(childHash) {
+			return 1, child.Edge.Path.len, nil
+		} else {
+			return 0, 0, errors.New("can't determine the child hash from the parent and child")
+		}
+	}
+
+	return 0, 1, nil
+}
+
+func assignChild(i, compressedParent int, parentNode *Node,
+	nilKey, leafKey, parentKey *Key, proofNodes []ProofNode, hashF hashFunc,
+) (*Key, error) {
+	childInd := i + compressedParent + 1
+	childKey, err := getChildKey(childInd, parentKey, leafKey, nilKey, proofNodes, hashF)
+	if err != nil {
+		return nil, err
+	}
+	if leafKey.Test(leafKey.len - parentKey.len - 1) {
+		parentNode.Right = childKey
+		parentNode.Left = nilKey
+	} else {
+		parentNode.Right = nilKey
+		parentNode.Left = childKey
+	}
+	return childKey, nil
+}
+
+// ProofToPath returns a set of storage nodes from the root to the end of the proof path.
+// It will contain the hashes of the children along the path, but only the key of the children
+// along the path. The final node must contain the hash of the leaf if the leaf is set.
+// It will not contain the leaf node even if it is set.
+func ProofToPath(proofNodes []ProofNode, leafKey *Key, hashF hashFunc) ([]StorageNode, error) {
```

Comment:
The code is much more readable 🎉  Great job!

Now, I cannot understand what is written in the fn's comment.
So we are given `proofNodes` and constructing `storageNodes` following along the path  to the single leaf node specified by given `leafKey`.
These `storageNodes` will have hashes of the children and keys? - not get this  `"but only the key of the children along the path"`.

Also: `"The final node must contain the hash of the leaf ..."` - which node is the final node, root or the nodes at the bottom of the trie 🤔 

And: `"It will not contain the leaf node even if it is set."`
Hmm, the leaf is set because we have the `leafKey` right? Also the leaf node refers to the `storageNodes` we are constructing here.

Can you please try to improve this comment?

Reply:
> These storageNodes will have hashes of the children and keys? - not get this "but only the key of the children along the path".

The storageNodes will have the hashes of the children, but only the key of the child along the path outlined by the proof.

> Also: "The final node must contain the hash of the leaf ..." - which node is the final node, root or the nodes at the bottom of the trie 🤔

Ah sorry, this is a confusing sentence. I was just trying to comment that we don't include the leaf node in the generated storage nodes. But if the path outlined by the proof leads the whole way to the leaf, and the leaf is set, then we should include it's hash.

> Hmm, the leaf is set because we have the leafKey right?

Nope. The Key that the proof node points to may not be set, but we can still have the key.

>  Also the leaf node refers to the storageNodes we are constructing here.

Yup

Reply:
I'll update the comment